lib/common/rules.c

```diff
@@ -1155,41 +1155,66 @@ pcmk__evaluate_attr_expression(const xmlNode *expression,
                               const pcmk_rule_input_t *rule_input)
 {
-    const char *class = crm_element_value(expr, PCMK_XA_CLASS);
-    const char *provider = crm_element_value(expr, PCMK_XA_PROVIDER);
-    const char *type = crm_element_value(expr, PCMK_XA_TYPE);
+    const char *id = NULL;
+    const char *value = NULL;
 
-    crm_trace("Testing rsc_defaults expression: %s", pcmk__xe_id(expr));
+    if ((rsc_expression == NULL) || (rule_input == NULL)) {
+        return EINVAL;
+    }
+
+    // Validate XML ID
+    id = pcmk__xe_id(rsc_expression);
+    if (pcmk__str_empty(id)) {
+        // Not possible with schema validation enabled
+        /* @COMPAT When we can break behavioral backward compatibility,
+         * fail the expression
+         */
+        pcmk__config_warn(PCMK_XE_RSC_EXPRESSION " has no " PCMK_XA_ID);
+        id = "without ID"; // for logging
+    }
 
-    if (class != NULL &&
-        !pcmk__str_eq(class, rule_input->rsc_standard, pcmk__str_none)) {
-        crm_trace("Class doesn't match: %s != %s", class, rule_input->rsc_standard);
+    // Compare resource standard
+    value = crm_element_value(rsc_expression, PCMK_XA_CLASS);
+    if (!pcmk__str_empty(value)
+        && !pcmk__str_eq(value, rule_input->rsc_standard, pcmk__str_none)) {
+        crm_trace(PCMK_XE_RSC_EXPRESSION " %s fails because "
+                  "standard doesn't match: %s != %s",
+                  id, value, pcmk__s(rule_input->rsc_standard, "unspecified"));
         return pcmk_rc_op_unsatisfied;
     }
 
-    if ((provider == NULL && rule_input->rsc_provider != NULL) ||
-        (provider != NULL && rule_input->rsc_provider == NULL) ||
-        !pcmk__str_eq(provider, rule_input->rsc_provider, pcmk__str_none)) {
-        crm_trace("Provider doesn't match: %s != %s", provider, rule_input->rsc_provider);
+    // Compare resource provider
+    value = crm_element_value(rsc_expression, PCMK_XA_PROVIDER);
+    if (!pcmk__str_empty(value)
+        && !pcmk__str_eq(value, rule_input->rsc_provider, pcmk__str_none)) {
+        crm_trace(PCMK_XE_RSC_EXPRESSION " %s fails because "
+                  "provider doesn't match: %s != %s",
+                  id, value, pcmk__s(rule_input->rsc_provider, "unspecified"));
         return pcmk_rc_op_unsatisfied;
     }
 
-    if (type != NULL &&
-        !pcmk__str_eq(type, rule_input->rsc_agent, pcmk__str_none)) {
-        crm_trace("Agent doesn't match: %s != %s", type, rule_input->rsc_agent);
+    // Compare resource agent type
+    value = crm_element_value(rsc_expression, PCMK_XA_TYPE);
+    if (!pcmk__str_empty(value)
+        && !pcmk__str_eq(value, rule_input->rsc_agent, pcmk__str_none)) {
+        crm_trace(PCMK_XE_RSC_EXPRESSION " %s fails because "
+                  "agent doesn't match: %s != %s",
+                  id, value, pcmk__s(rule_input->rsc_agent, "unspecified"));
         return pcmk_rc_op_unsatisfied;
     }
 
+    crm_trace(PCMK_XE_RSC_EXPRESSION " %s passes (agent %s)",
```

Comment:
I assume this is for convenience of reusing `value`. However, it is kind of odd if we have a rule that specifies standard and provider and we don't mention them here. We do give the rule ID so a reader can go look it up, but the agent name is no more special than the standard and provider names.

Up to you, it's just tracing

---

Also surprised to see that a totally empty resource expression is valid (no ID, no standard, no provider, no agent).

Maybe one more test for "empty resource expression matches"? All the individual missing cases are covered.

Reply:
Yes, I didn't think it was worth separate variables just to trace them, but might as well

Reply:
It probably would have been a better design to require at least one of class, provider, and type, but it doesn't matter enough to change now